### PR TITLE
Add `--all` flag to `cargo test`

### PR DESCRIPTION
**Pull request by euclio**
Work towards #2878.

This flag allows a user to test all members of a workspace by using `cargo test --all`. The command is also supported when using a virtual workspace manifest.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! Unfortunately I'll be pretty busy this week so it may take awhile to get back to this, but I haven't forgotten it!

File: src/cargo/core/resolver/encode.rs [hunk 1]
```diff
@@ -281,14 +281,15 @@ impl<'a, 'cfg> Encodable for WorkspaceResolve<'a, 'cfg> {
 
         let root = self.ws.members().max_by_key(|member| {
             member.name()
-        }).unwrap().package_id();
+        }).map(Package::package_id);
```
Comment (alexcrichton) [inline src/cargo/core/resolver/encode.rs]: Hm were the changes in this file required to get tests to pass? Or just cleaning things up?

File: src/cargo/ops/cargo_compile.rs [hunk 2]
```diff
@@ -169,20 +178,44 @@ pub fn compile_ws<'a>(ws: &Workspace<'a>,
...
-    let profiles = root_package.manifest().profiles();
-    if spec.len() == 0 {
-        try!(generate_targets(root_package, profiles, mode, filter, release));
-    }
+    let spec: Cow<'a, [String]> = match spec {
+        Packages::Packages(spec) => spec.into(),
+        Packages::All => ws.members()
+                           .map(|package| {
+                               let package_id = package.package_id();
+                               PackageIdSpec::from_package_id(package_id).to_string()
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: Instead of round-tripping through `PackageIdSpec`, perhaps the resolution to a list of local packages could be done sooner?

File: src/cargo/ops/cargo_rustc/context.rs [hunk 3]
```diff
@@ -28,7 +28,7 @@ pub struct Unit<'a> {
 pub struct Context<'a, 'cfg: 'a> {
     pub config: &'cfg Config,
     pub resolve: &'a Resolve,
-    pub current_package: PackageId,
+    pub current_package: Option<PackageId>,
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/context.rs]: Hm so I think this is a code smell we need to clean up (not your fault, a historical one). The concept of a "current package" is basically no longer valid in today's world of workspaces and such. Could this field get outright removed as part of this change?

File: src/cargo/ops/cargo_compile.rs [hunk 4]
```diff
@@ -169,20 +178,44 @@ pub fn compile_ws<'a>(ws: &Workspace<'a>,
...
     if spec.len() > 0 {
-        for p in spec {
+        for ref p in spec.iter() {
             pkgids.push(try!(resolve_with_overrides.query(&p)));
         }
     } else {
+        let root_package = try!(ws.current());
+        if spec.len() == 0 {
+            try!(generate_targets(root_package, &profiles, mode, filter, release));
+        }
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: This kinda sent me spinning for a bit. I think this `if` statement isn't necessary, right? It's implied to always be true given the above condition.

File: src/cargo/ops/cargo_compile.rs [hunk 5]
```diff
@@ -169,20 +178,44 @@ pub fn compile_ws<'a>(ws: &Workspace<'a>,
...
+            release: Profile::default_release(),
+            dev: Profile::default_dev(),
+            test: Profile::default_test(),
+            test_deps: Profile::default_dev(),
+            bench: Profile::default_bench(),
+            bench_deps: Profile::default_release(),
+            doc: Profile::default_doc(),
+            custom_build: Profile::default_custom_build(),
+        },
+    };
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: Hm this is somewhat worrisome to me. This is where I think we'll want to learn about the profiles from the workspace root crate instead of whatever the current crate happens to be.

File: src/cargo/core/resolver/encode.rs (hunk 1 above)
Comment (euclio) [inline src/cargo/core/resolver/encode.rs]: Yes, that panic is triggered when running `test` on a virtual manifest. It was currently prevented by the "no targets specified in the manifest" error.

File: src/cargo/ops/cargo_compile.rs (hunk 4 above)
Comment (euclio) [inline src/cargo/ops/cargo_compile.rs]: Fixed.

File: src/cargo/ops/cargo_rustc/context.rs (hunk 3 above)
Comment (euclio) [inline src/cargo/ops/cargo_rustc/context.rs]: I can try that, sure.

File: src/cargo/ops/cargo_compile.rs (hunk 2 above)
Comment (euclio) [inline src/cargo/ops/cargo_compile.rs]: Where would be the right place to do that? I didn't want to place it too high up so that future `--all` commands could share this logic.

File: src/cargo/ops/cargo_rustc/context.rs (hunk 3 above)
Comment (euclio) [inline src/cargo/ops/cargo_rustc/context.rs]: I think I need some clarification. What does "current package" mean in this context? Isn't it the package of the directory that we issue the compile in? Why is that not needed?

File: src/cargo/ops/cargo_compile.rs (hunk 5 above)
Comment (matklad) [inline src/cargo/ops/cargo_compile.rs]: Not exactly yet, see #3249 :)

Comment (euclio) [inline src/cargo/ops/cargo_compile.rs]: @matklad Awesome! I can rebase on top of your PR.

File: src/cargo/core/resolver/encode.rs (hunk 1 above)
Comment (alexcrichton) [inline src/cargo/core/resolver/encode.rs]: Ok, sounds good to me

File: src/cargo/ops/cargo_compile.rs (hunk 2 above)
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: Maybe the concept of "all" could be plumbed farther down as an enum?

File: src/cargo/ops/cargo_compile.rs (hunk 5 above)
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: Ah yeah #3249 is exactly what I had in mind here

File: src/cargo/ops/cargo_rustc/context.rs (hunk 3 above)
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/context.rs]: Yeah the intention is there really is not right question to what the "current package" means. It has different interpretations depending on where you are asking it from, so that's why I figure it'd be best to remove and we can answer that question on a case-by-case basis.

**Comment by euclio** [comment]
Rebased. Hopefully will get a chance to work on this more later this week.

File: src/cargo/ops/cargo_compile.rs (hunk 2 above)
Comment (euclio) [inline src/cargo/ops/cargo_compile.rs]: Plumbed down where? It makes sense to me to pass `Packages` down through `resolve_dependencies` but don't you still need to get the `PackageIdSpec`s of the workspace here to resolve the targets?

File: src/cargo/ops/cargo_compile.rs [hunk 6]
```diff
@@ -92,8 +99,10 @@ pub enum CompileFilter<'a> {
 
 pub fn compile<'a>(ws: &Workspace<'a>, options: &CompileOptions<'a>)
                    -> CargoResult<ops::Compilation<'a>> {
-    for key in try!(ws.current()).manifest().warnings().iter() {
-        try!(options.config.shell().warn(key))
+    if let Some(root_package) = ws.current_opt() {
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: Random drive-by comment, but could this actually iterate over all members of the workspace? That's the intention here, at least

File: src/cargo/ops/cargo_compile.rs (hunk 2 above)
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: Yeah let's pass this down to `resolve_dependencies`, and perhaps that can also return a list of specs?

**Comment by euclio** [comment]
Rebased, and ready for another round of review, @alexcrichton

File: src/cargo/ops/cargo_compile.rs [hunk 7]
```diff
@@ -167,27 +187,23 @@ pub fn compile_ws<'a>(ws: &Workspace<'a>,
...
     let mut pkgids = Vec::new();
     if spec.len() > 0 {
-        for p in spec {
-            pkgids.push(resolve_with_overrides.query(&p)?);
+        for p in spec.iter() {
+            pkgids.push(resolve_with_overrides.query(&p.to_string())?);
         }
     } else {
+        let root_package = ws.current()?;
+        generate_targets(root_package, profiles, mode, filter, release)?;
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_compile.rs]: I think this was special cased above to get these errors *before* the resolution process happened (which may involve downloads and such). Perhaps that logic could still be special cased?

File: src/cargo/ops/cargo_rustc/context.rs [hunk 8]
```diff
@@ -803,8 +803,9 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
     }
 
     pub fn show_warnings(&self, pkg: &PackageId) -> bool {
-        pkg == &self.current_package || pkg.source_id().is_path() ||
-            self.config.extra_verbose()
+        self.ws.current_opt().map_or(false, |p| *pkg == *p.package_id())
+            || pkg.source_id().is_path()
+            || self.config.extra_verbose()
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/context.rs]: I think this can be updated to just the clause:

```rust
pkg.source_id().is_path() || self.config.extra_verbose()
```

File: src/cargo/ops/cargo_rustc/context.rs [hunk 9]
```diff
@@ -417,7 +416,8 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         // we don't want to link it up.
         if src_dir.ends_with("deps") {
             // Don't lift up library dependencies
-            if unit.pkg.package_id() != &self.current_package && !unit.target.is_bin() {
+            if self.ws.current_opt().map_or(false, |p| unit.pkg.package_id() != p.package_id())
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/context.rs]: I think here we can actually just test:

```rust
if !unit.pkg.package_id().source_id().is_path() && !unit.target.is_bin() {
    // ...
}
```

That is, all path dependencies should hit the code path of `Some` below

File: src/cargo/ops/cargo_rustc/context.rs (hunk 8 above)
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/context.rs]: Well technically actually, we should show warnings for all *workspace members*, not not all path dependencies.

File: src/cargo/ops/cargo_rustc/context.rs (hunk 9 above)
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/context.rs]: er actually, we should replace the `current_package` check with a check if the crate is in the workspace rather than all path dependencies.

File: src/cargo/ops/cargo_rustc/context.rs [hunk 10]
```diff
@@ -273,7 +272,7 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
 
     /// Returns the appropriate directory layout for either a plugin or not.
     pub fn layout(&self, unit: &Unit) -> LayoutProxy {
-        let primary = unit.pkg.package_id() == &self.current_package;
+        let primary = self.ws.current_opt().map_or(false, |p| unit.pkg.package_id() == p.package_id());
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/context.rs]: I think this can be replaced with:

```rust
let primary = is_member_of_current_workspace(self, unit);
```

File: src/cargo/ops/cargo_rustc/job_queue.rs [hunk 11]
```diff
@@ -198,7 +198,9 @@ impl<'a> JobQueue<'a> {
         }
 
         let build_type = if self.is_release { "release" } else { "debug" };
-        let profile = cx.lib_profile(&cx.current_package);
+        let profile = cx.ws.current_opt().map_or_else(Profile::default, |p| {
+            cx.lib_profile(p.package_id()).to_owned()
+        });
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/job_queue.rs]: The package argument here to `lib_profile` is actually ignored (long story), so let's just go ahead and remove that argument and then we can avoid the workaround here.

File: src/cargo/ops/cargo_rustc/mod.rs [hunk 12]
```diff
@@ -545,7 +545,9 @@ fn build_base_args(cx: &Context,
     let prefer_dynamic = (unit.target.for_host() &&
                           !unit.target.is_custom_build()) ||
                          (crate_types.contains(&"dylib") &&
-                          unit.pkg.package_id() != &cx.current_package);
+                          cx.ws.current_opt().map_or(false, |p| {
+                              *p.package_id() != *unit.pkg.package_id()
+                          }));
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/mod.rs]: Wow I have absolutely no clue what this clause is doing (the old one).

Let's just replace it though with, like above, a test whether the package is in the current workspace.

**Comment by euclio** [comment]
Still working on this, haven't had much time for open source lately! Hoping to finish up by the weekend.

**Comment by alexcrichton** [comment]
Ah no worries, let me know if you need any help getting it over the finish line!

File: src/cargo/ops/cargo_rustc/context.rs (hunk 9 above)
Comment (euclio) [inline src/cargo/ops/cargo_rustc/context.rs]: What's the correct way to check this?

`ws.members().find(|&p| p == unit.pkg).is_some()` causes test failures.

File: src/cargo/ops/cargo_rustc/mod.rs (hunk 12 above)
Comment (euclio) [inline src/cargo/ops/cargo_rustc/mod.rs]: Same as above.

**Comment by euclio** [comment]
@alexcrichton Most comments addressed, I commented on the remaining issues.

**Comment by alexcrichton** [comment]
Sorry for taking awhile to get back to this! I just want to dig into the "is a member" issue a bit and then we should be able to land this.

File: src/cargo/ops/cargo_rustc/mod.rs (hunk 12 above)
Comment (euclio) [inline src/cargo/ops/cargo_rustc/mod.rs]: Ditto.

**Comment by euclio** [comment]
@alexcrichton, no problem! Everything should be addressed now.

**Comment by alexcrichton** [comment]
@bors: r+

Awesome, thanks!

**Comment by khuey** [comment]
Was it expected that this changed cargo's behavior so that tests of dependent lib crates are no longer dropped in the `target/debug` folder?

**Comment by alexcrichton** [comment]
@khuey no I think that may have been accidental! Oddly enough https://github.com/rust-lang/cargo/issues/3660 was just reported as well and I believe that bug was fixed in https://github.com/rust-lang/cargo/pull/3478, could you test beta/nightly to see if they're working for you?

**Comment by khuey** [comment]
It's still broken on tip (de2919fd32f1d4f599bb59ac78ea4f01f3a33e46)

**Comment by khuey** [comment]
Woops, I was testing the wrong binary.  It is indeed fixed on tip.  I'll bisect for a fix cset.

**Comment by alexcrichton** [comment]
Ok, thanks for checking!

**Comment by khuey** [comment]
It was indeed fixed by 3cbf141.

The regression is present in the cargo that ships with rustc 1.15.  That's not a big deal for me (I just won't use it) but something you may want to be aware of.

**Comment by alexcrichton** [comment]
Sure thing, thanks for the heads up regardless
